dialogue-core/src/main/java/com/palantir/dialogue/core/StickyConcurrencyLimitedChannel.java

```diff
@@ -0,0 +1,110 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.core.CautiousIncreaseAggressiveDecreaseConcurrencyLimiter.Behavior;
+import com.palantir.dialogue.futures.DialogueFutures;
+import com.palantir.logsafe.SafeArg;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class StickyConcurrencyLimitedChannel implements LimitedChannel {
+
+    private static final Logger log = LoggerFactory.getLogger(StickyConcurrencyLimitedChannel.class);
+
+    private final NeverThrowLimitedChannel delegate;
+    private final CautiousIncreaseAggressiveDecreaseConcurrencyLimiter limiter;
+    private final String channelNameForLogging;
+
+    @VisibleForTesting
+    StickyConcurrencyLimitedChannel(
+            LimitedChannel delegate,
+            CautiousIncreaseAggressiveDecreaseConcurrencyLimiter limiter,
+            String channelNameForLogging) {
+        this.delegate = new NeverThrowLimitedChannel(delegate);
+        this.limiter = limiter;
+        this.channelNameForLogging = channelNameForLogging;
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(
+            Endpoint endpoint, Request request, LimitEnforcement limitEnforcement) {
+        Optional<CautiousIncreaseAggressiveDecreaseConcurrencyLimiter.Permit> maybePermit =
+                limiter.acquire(limitEnforcement);
+        if (maybePermit.isPresent()) {
+            CautiousIncreaseAggressiveDecreaseConcurrencyLimiter.Permit permit = maybePermit.get();
+            logPermitAcquired();
+
+            // This is a trade-off to solve an edge case where the first request on this channel could get
+            // concurrency limited: we would then have to rely on another request coming in to try scheduling the
+            // first request again. If that request never came, we would lock up.
+            // To combat that we have a way to instruct downstream channel to ignore its capacity limits,
+            // and let the request through.
+            Optional<ListenableFuture<Response>> result = delegate.maybeExecute(
+                    endpoint,
+                    request,
+                    permit.isOnlyInFlight() || limitEnforcement == LimitEnforcement.DANGEROUS_BYPASS_LIMITS
+                            ? LimitEnforcement.DANGEROUS_BYPASS_LIMITS
+                            : LimitEnforcement.DEFAULT_ENABLED);
```

Comment:
I think this is easier to understand with a slight refactor:
```suggestion
                    permit.isOnlyInFlight()
                            ? LimitEnforcement.DANGEROUS_BYPASS_LIMITS
                            : limitEnforcement);
```

Reply:
Much better!